File: UnitTests/BraintreeCoreTests/Helpers/BTHTTPTestProtocol.m
#import "BTHTTPTestProtocol.h"
@import BraintreeCore;
@import BraintreeCoreSwift;

@implementation BTHTTPTestProtocol

+ (BOOL)canInitWithRequest:(__unused NSURLRequest *)request {
    return YES;
}

+ (NSURLRequest *)canonicalRequestForRequest:(NSURLRequest *)request {
    return request;
}

- (void)startLoading {
    id<NSURLProtocolClient> client = self.client;

    NSHTTPURLResponse *response = [[NSHTTPURLResponse alloc] initWithURL:self.request.URL
                                                              statusCode:200
                                                             HTTPVersion:@"HTTP/1.1"
                                                            headerFields:@{@"Content-Type": @"application/json"}];

    NSData *archivedRequest = [NSKeyedArchiver archivedDataWithRootObject:self.request requiringSecureCoding:YES error:nil];
    NSString *base64ArchivedRequest = [archivedRequest base64EncodedStringWithOptions:0];

    NSData *requestBodyData;
    if (self.request.HTTPBodyStream) {
        NSInputStream *inputStream = self.request.HTTPBodyStream;
        [inputStream open];
        NSMutableData *mutableBodyData = [NSMutableData data];

        while ([inputStream hasBytesAvailable]) {
            uint8_t buffer[128];
            NSUInteger bytesRead = [inputStream read:buffer maxLength:128];
            [mutableBodyData appendBytes:buffer length:bytesRead];
        }
        [inputStream close];
        requestBodyData = [mutableBodyData copy];
    } else {
        requestBodyData = self.request.HTTPBody;
    }

    NSDictionary *responseBody = @{ @"request": base64ArchivedRequest,
                                    @"requestBody": [[NSString alloc] initWithData:requestBodyData encoding:NSUTF8StringEncoding] };

    [client URLProtocol:self didReceiveResponse:response cacheStoragePolicy:NSURLCacheStorageNotAllowed];

    [client URLProtocol:self didLoadData:[NSJSONSerialization dataWithJSONObject:responseBody options:NSJSONWritingPrettyPrinted error:NULL]];

    [client URLProtocolDidFinishLoading:self];
}

- (void)stopLoading {
}

+ (NSURL *)testBaseURL {
    NSURLComponents *components = [[NSURLComponents alloc] init];
    components.scheme = kBTHTTPTestProtocolScheme;
    components.host = kBTHTTPTestProtocolHost;
    components.path = kBTHTTPTestProtocolBasePath;
    components.port = kBTHTTPTestProtocolPort;
    return components.URL;
}

+ (NSURLRequest *)parseRequestFromTestResponseBody:(BTJSON *)responseBody {
//    return [NSKeyedUnarchiver unarchiveObjectWithData:[[NSData alloc] initWithBase64EncodedString:[responseBody[@"request"] asString] options:0]];
    NSData *dataToUnarchive = [[NSData alloc] initWithBase64EncodedString:[responseBody[@"request"] asString] options:0];
    return [NSKeyedUnarchiver unarchivedObjectOfClass:NSURLRequest.class fromData:dataToUnarchive error:nil];
}

+ (NSString *)parseRequestBodyFromTestResponseBody:(BTJSON *)responseBody {
    return [responseBody[@"requestBody"] asString];
}

@end

File: UnitTests/BraintreeCoreTests/Helpers/BTHTTPTestProtocol.swift
import Foundation
import BraintreeCore

@objcMembers class BTHTTPTestProtocol: URLProtocol {

    override class func canInit(with request: URLRequest) -> Bool {
        true
    }

    override class func canonicalRequest(for request: URLRequest) -> URLRequest {
        request
    }

    override func startLoading() {
        guard let client = client else { return }

        let response = HTTPURLResponse(
            url: request.url!,
            statusCode: 200,
            httpVersion: "HTTP/1.1",
            headerFields: ["Content-Type": "application/json"]
        )

        let archivedRequest: Data = try! NSKeyedArchiver.archivedData(withRootObject: request, requiringSecureCoding: true)
        let base64ArchivedRequest = archivedRequest.base64EncodedString()
        var requestBodyData: Data? = nil

        if request.httpBodyStream != nil {
            guard let inputStream = request.httpBodyStream else { return }
            inputStream.open()
            var mutableBodyData = Data()

            while inputStream.hasBytesAvailable {
                let bufferSize: Int = 128
                let buffer = UnsafeMutablePointer<UInt8>.allocate(capacity: bufferSize)
                let bytesRead = inputStream.read(buffer, maxLength: bufferSize)
                mutableBodyData.append(buffer, count: bytesRead)
            }
            inputStream.close()
            requestBodyData = mutableBodyData
        } else {
            requestBodyData = request.httpBody ?? Data()
        }

        let responseBody = [
            "request": base64ArchivedRequest,
            "requestBody": String(data: requestBodyData ?? Data(), encoding: .utf8)
        ]

        client.urlProtocol(self, didReceive: response!, cacheStoragePolicy: URLCache.StoragePolicy.notAllowed)
        client.urlProtocol(self, didLoad: try! JSONSerialization.data(withJSONObject: responseBody, options: .prettyPrinted))
        client.urlProtocolDidFinishLoading(self)
    }

    override func stopLoading() {
        // no-op
    }

    static func testBaseURL() -> URL {
        var components = URLComponents()
        components.scheme = "bt-http-test"
        components.host = "base.example.com"
        components.path = "/base/path"
        components.port = 1234
        return components.url!
    }

    static func parseRequestFromTestResponseBody(_ responseBody: BTJSON) -> URLRequest {
        let dataToUnarchive = Data(base64Encoded: responseBody["request"].asString() ?? "") ?? Data()
        let urlRequest = try? NSKeyedUnarchiver.unarchivedObject(ofClass: NSURLRequest.self, from: dataToUnarchive)! as URLRequest
        return urlRequest ?? URLRequest(url: URL(string: "")!)
    }

    static func parseRequestBodyFromTestResponseBody(_ responseBody: BTJSON) -> String {
        responseBody["requestBody"].asString() ?? ""
    }
}
